started working on client push methods

foxriver76 opened the pull request:
- will close #2373

foxriver76 commented on packages/adapter/src/lib/adapter/userInterfaceMessagingController.ts:
```diff
@@ -0,0 +1,184 @@
+
+        return states.pushMessage(handler.from, {
+            command: 'im',
+            message: { m: handler.type, s: handler.sid, d: data },
```
Can we make the property names and the command more meaningful @GermanBluefox? Maybe I also misinterpreted, that commadn needs to be `im` like in the example at https://github.com/ioBroker/ioBroker.socket-classes#gui-subscribes

GermanBluefox replied:
No. I save here the traffic. Because I send with it the images to GUI and additional 100 bytes in every frame could be easy transformed into 1k by 10 frames per second.

GermanBluefox replied:
And exactly because of these short names I want to hide it into adapter.js, so the dev does not use these short names

foxriver76 replied:
Okay, if you say it mattered in your tests we can keep it

GermanBluefox commented:
LGTM

AlCalzone commented on packages/adapter/src/lib/_Types.ts:
```diff
@@ -46,6 +50,65 @@ export interface AdapterOptions {
+    (
+        | {
+              /** Reason for unsubscribe */
+              reason: Exclude<UserInterfaceClientUnsubscribeReason, ClientUnsubscribeReason>;
```
```suggestion
              reason: Exclude<UserInterfaceClientUnsubscribeReason, ClientUnsubscribeReason>;
              message?: undefined;
```
avoids having to test `message in something` before accessing `message`

AlCalzone reviewed (changes requested):
I didn't see how the client subscribes to these messages. Would be cool if there was an easy way to give the data a type without resorting to `as Foobar` in every client handler.

Something like
```ts
subscribe<MyData>((data) => {
   // data is MyData
})
```

Also, do we really want to call the method
`sendToUserInterfaceClient`? That's quite a mouthful to type. Suggestions in order of my personal preference:
- `pushToUI`
- `sendToUI`
- `sendToUIClient`

foxriver76 commented:
Thanks for the review, I have addressed the comments.

Regarding making the subscribe generic, I don't know how this will bring any benefit, as we are just managing subscribes here. Not receiving any relevant data for the backend normally. The type of `data` will be more relevant on the frontend side, as this is what the backend pushes. The `data: unknown` will never be part of a subscribe handler here.

AlCalzone commented:
> The type of `data` will be more relevant on the frontend side, as this is what the backend pushes

That's what I meant - not part of this PR I guess.
